### PR TITLE
[IMP] base: index optimization on ir.translation

### PR DESCRIPTION
**Pull request by mart-e**
Since postgresql 10.0, hash indexes are safe to be used
https://www.postgresql.org/docs/10/release-10.html#id-1.11.6.17.5.3.3

> - Add write-ahead logging support to hash indexes (Amit Kapila)
>   - This makes hash indexes crash-safe and replicatable. The former warning message about their use is removed.
> - Improve hash index performance (Amit Kapila, Mithun Cy, Ashutosh Sharma)

Hash indexes [are faster](http://amitkapila16.blogspot.com/2017/03/hash-indexes-are-faster-than-btree.html) but only works on `=` operator. So ideal for translations where we do exact match searches.

A standard ir_translation table with 138k records (67 modules installed and 7 languages activated), after a vacuum analyze.
```
 table_size | indexes_size | total_size 
------------+--------------+------------
 24 MB      | 59 MB        | 71 MB
```

With this branch, same modules and languages:

```
 table_size | indexes_size | total_size 
------------+--------------+------------
 24 MB      | 47 MB        | 71 MB
```

Remove index on `res_id` and `comments` as not useful.

Switching from `md5` index to native hash index reduced the size of the index from 11MB to 5MB.
Switching `ir_translation_unique` to a partial index reduced the size of the index from 21MB to 8MB.

Unique hash index is not (yet?) possible so need to maintain the other `md5` unique indexes.

Remove the `order='module'`, we don't care of the order in this specific case.

PS: to extrapolate, on odoo.com, the ir_translation takes 380MB for data and 2023MB for indexes.

### CODE REVIEW

File: odoo/addons/base/models/ir_translation.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ class IrTranslation(models.Model):
     _log_access = False
 
     name = fields.Char(string='Translated field', required=True)
-    res_id = fields.Integer(string='Record ID', index=True)
+    res_id = fields.Integer(string='Record ID')
```
Comment (odony) [inline odoo/addons/base/models/ir_translation.py]: It does seem to be useful in some cases. Some usage statistics in a real database:

```
                 index                  |     table      |  idx_scan  | pg_size_pretty 
----------------------------------------+----------------+------------+----------------
 ir_translation_unique                  | ir_translation |  344815142 | 336 MB
 ir_translation_ltn                     | ir_translation |      96276 | 209 MB
 ir_translation_name_lang_res_id_id_idx | ir_translation | 1524412718 | 192 MB
 ir_translation_type_name_res_id_idx    | ir_translation |   87583075 | 179 MB
 ir_translation_module_index            | ir_translation |  243636669 | 167 MB
 ir_translation_src_md5                 | ir_translation |   12213626 | 164 MB
 ir_translation_type_index              | ir_translation |     151163 | 162 MB
 ir_translation_comments_pattern_idx    | ir_translation |          0 | 156 MB
 ir_translation_comments_index          | ir_translation |          0 | 156 MB
 ir_translation_pkey                    | ir_translation | 4755952486 | 134 MB
 ir_translation_res_id_index            | ir_translation |  294367430 | 90 MB
 ir_translation_model_unique            | ir_translation |     111931 | 85 MB
 ir_translation_code_unique             | ir_translation |     129030 | 7480 kB
```

We can however see that the index on `comments` is unused, even though we search on this column in `get_translations_for_webclient()`.

**Comment by mart-e** [comment]
> We can however see that the index on `comments` is unused, even though we search on this column in `get_translations_for_webclient()`.

The `ir_translation_module_index` seems enough to be used [for these queries](https://explain.depesz.com/s/QJj).

What is odd, is that the `ir_translation_model_unique` is not used so much, For instance, when reading a product, we join on `ir.translation` but the partial index `ir_translation_model_unique` is not used (even if was supposed to match perfectly).
 - [On standard db](https://explain.depesz.com/s/R1dt) -> using `ir_translation_unique` instead
- [With this branch](https://explain.depesz.com/s/r3gs) -> using `ir_translation_res_id_index` instead
- [With custom index](https://explain.depesz.com/s/UgW) -> using `ir_translation_name_lang_res_id_id_idx`

While these partial indexes help us for the import (thanks to `unique`), it doesn't seem to be so benefical for retrieval...

**Comment by odony** [comment]
> > We can however see that the index on `comments` is unused, even though we search on this column in `get_translations_for_webclient()`.
> 
> The `ir_translation_module_index` seems enough to be used [for these queries](https://explain.depesz.com/s/QJj).

In fact the index was not used in this case because `unaccent` is installed and the query did not match the index.
But creating a proper index does speed up the loading of web translations 10x (10ms instead of 150ms):

```sql
CREATE INDEX ON ir_translation (module, lang, unaccent(comments))
          WHERE value != '' AND value IS NOT NULL;
```

Judging by the activity, it's an important index to add. Added 30 minutes ago and there are already 700k index scans!

**Comment by mart-e** [comment]
> In fact the index was not used in this case because `unaccent` is installed and the query did not match the index.

Do you suggest we add an unaccent variation of the index if the `unaccent` is activated?
By the way, using the config parameter may not be a great indicator, cf #46534

**Comment by xmo-odoo** [comment]
> Unique hash index is not (yet?) possible so need to maintain the other md5 unique indexes.

Apparently it's kinda possible using exclusion constraints instead:
```sql
alter table <table> add constraint <constraint> exclude using hash (<field> with =);
```
It's unclear whether we'd also need to create a separate hash index, the description of this constraint is listed in the INDEXES section in psql so I'd guess not but that should probably be query-planned.

> Do you suggest we add an unaccent variation of the index if the unaccent is activated?

It doesn't seem to be a standard index though?

**Comment by nilshamerlinck** [comment]
Hi, do you use a specific version of `unaccent` in your deployments? afaik the standard one cannot be used in indexes:

```
> ERROR:  functions in index expression must be marked IMMUTABLE
```

**Comment by yajo** [comment]
The answer to https://github.com/odoo/odoo/pull/45958#issuecomment-606614247 is that Odoo alters the unaccent function to be considered immutable: https://github.com/odoo/odoo/blob/8e5dc5352c80314db3b60e3a94fd343f96828f41/odoo/service/db.py#L124-L132

See commit message in 4087bcdc5e4dde3fc7dc1eb2d01cd1cdd55ae021 for more explanations.

**Pull request closed without merging**